packages/sources/coinpaprika-state/src/transport/coinpaprika-state.ts

```diff
@@ -0,0 +1,238 @@
+      apiEndpoint: context.adapterSettings.API_ENDPOINT,
+      apiKey: context.adapterSettings.API_KEY,
+      requestTimeoutMs: context.adapterSettings.REQUEST_TIMEOUT_MS,
+    }
+
+    const callbacks: SSEConnectionCallbacks = {
+      onData: async (eventType: string, data: string) => {
+        await this.handleSSEEvent(eventType, data)
+      },
+      onError: (error: Error) => {
+        logger.error(`SSE stream error: ${error.message}`)
+      },
+      onConnectionError: async (status: number) => {
+        await this.handleConnectionError(status)
+      },
+      onReconnectNeeded: () => {
+        logger.info('SSE ended; will reconnect on next cycle')
+      },
+    }
+
+    await this.connectionManager.connect(pairRequests, config, callbacks)
+  }
+
+  private async handleConnectionError(status: number): Promise<void> {
+    const mappedStatus = status === 500 ? 502 : status
+    const errorMessage = `HTTP ${status} error from provider`
+
+    const errorResponse: AdapterResponse<TransportTypes['Response']> = {
+      statusCode: mappedStatus,
+      errorMessage,
+      timestamps: {
+        providerDataRequestedUnixMs: Date.now(),
+        providerDataReceivedUnixMs: Date.now(),
+        providerIndicatedTimeUnixMs: undefined,
+      },
+    }
+
+    for (const params of this.activePairs.values()) {
+      await this.responseCache.write(this.name, [{ params, response: errorResponse }])
+    }
+  }
+
+  private async handleStreamError(errorData: string): Promise<void> {
+    const errorResponse: AdapterResponse<TransportTypes['Response']> = {
+      statusCode: 502,
+      errorMessage: `SSE stream error: ${errorData}`,
+      timestamps: {
+        providerDataRequestedUnixMs: Date.now(),
+        providerDataReceivedUnixMs: Date.now(),
+        providerIndicatedTimeUnixMs: undefined,
+      },
+    }
+
+    for (const params of this.activePairs.values()) {
+      await this.responseCache.write(this.name, [{ params, response: errorResponse }])
+    }
+  }
+
+  private async handleStreamData(param: RequestParams, streamData: CoinpaprikaStreamData) {
+    const statePrice = toNumber(streamData.state_price)
+    const blockTime = toNumber(streamData.block_time)
+
+    // guard against bad payloads
+    if (!Number.isFinite(statePrice) || !Number.isFinite(blockTime) || blockTime <= 0) {
+      logger.warn(
+        `Bad numeric fields for ${param.base}/${param.quote}: ${JSON.stringify(streamData)}`,
+      )
+      return
+    }
+
+    const response: AdapterResponse<TransportTypes['Response']> = {
+      statusCode: 200,
+      result: statePrice,
+      data: {
+        result: statePrice,
+      },
+      timestamps: {
+        providerDataRequestedUnixMs: Date.now(),
+        providerDataReceivedUnixMs: Date.now(),
+        providerIndicatedTimeUnixMs: blockTime * 1000,
+      },
+    }
+    logger.debug(`tick ${param.base}/${param.quote}=${statePrice} t=${blockTime}`)
+    logger.trace(JSON.stringify(response))
```

Comment:
These will stringify the response always regardless of whether the log is printed due to the selected level or not; if the stream is very heavy with updates, that's a lot of cpu time spent on this. We can either pass the object as-is and not stringify (the printing should be handled in the logger) or remove the log altogether